File: selenium/utils/excelReporter.js
const ExcelJS = require("exceljs");
const path = require("path");
const fs = require("fs");

class ExcelReporter {
  constructor(outputPath) {
    this.outputPath = outputPath;
    this.rows = [];
  }

  addResult({ testId, module, scenario, status, durationMs, details = "" }) {
    this.rows.push({
      testId,
      module,
      scenario,
      status: String(status).toUpperCase(),
      durationMs: Number(durationMs.toFixed(1)),
      details,
      executedAt: new Date().toISOString(),
    });
  }

  async write() {
    const workbook = new ExcelJS.Workbook();
    const summary = workbook.addWorksheet("Summary");
    const details = workbook.addWorksheet("Test Cases");
    const analysis = workbook.addWorksheet("Excel Analysis");

    const passed = this.rows.filter((r) => r.status === "PASSED").length;
    const failed = this.rows.filter((r) => r.status === "FAILED").length;
    const total = this.rows.length;
    const passRate = total ? Number(((passed / total) * 100).toFixed(2)) : 0;

    summary.addRow(["Metric", "Value"]);
    summary.addRow(["Total Test Cases", total]);
    summary.addRow(["Passed", passed]);
    summary.addRow(["Failed", failed]);
    summary.addRow(["Pass Rate (%)", passRate]);
    summary.addRow(["Generated At", new Date().toISOString()]);
    summary.addRow(["Framework", "Selenium Web E2E + Excel Analysis (Node.js)"]);
    this.#styleHeader(summary);

    details.addRow([
      "Test ID",
      "Module",
      "Scenario",
      "Status",
      "Duration (ms)",
      "Details",
      "Executed At",
    ]);
    for (const row of this.rows) {
      details.addRow([
        row.testId,
        row.module,
        row.scenario,
        row.status,
        row.durationMs,
        row.details,
        row.executedAt,
      ]);
    }
    this.#styleHeader(details);
    this.#colorStatus(details);

    analysis.addRow(["Module", "Total", "Passed", "Failed", "Pass Rate (%)"]);
    const modules = {};
    for (const row of this.rows) {
      if (!modules[row.module]) {
        modules[row.module] = { total: 0, passed: 0, failed: 0 };
      }
      modules[row.module].total += 1;
      if (row.status === "PASSED") modules[row.module].passed += 1;
      else modules[row.module].failed += 1;
    }
    for (const [module, stats] of Object.entries(modules).sort()) {
      const rate = stats.total
        ? Number(((stats.passed / stats.total) * 100).toFixed(2))
        : 0;
      analysis.addRow([module, stats.total, stats.passed, stats.failed, rate]);
    }
    this.#styleHeader(analysis);

    fs.mkdirSync(path.dirname(this.outputPath), { recursive: true });
    await workbook.xlsx.writeFile(this.outputPath);
    return this.outputPath;
  }

  #styleHeader(sheet) {
    const row = sheet.getRow(1);
    row.font = { bold: true, color: { argb: "FFFFFFFF" } };
    row.fill = {
      type: "pattern",
      pattern: "solid",
      fgColor: { argb: "FF1F4E79" },
    };
  }

  #colorStatus(sheet) {
    sheet.eachRow((row, rowNumber) => {
      if (rowNumber === 1) return;
      const cell = row.getCell(4);
      const value = String(cell.value || "").toUpperCase();
      if (value === "PASSED") {
        cell.fill = {
          type: "pattern",
          pattern: "solid",
          fgColor: { argb: "FFC6EFCE" },
        };
      } else if (value === "FAILED") {
        cell.fill = {
          type: "pattern",
          pattern: "solid",
          fgColor: { argb: "FFFFC7CE" },
        };
      }
    });
  }
}

module.exports = { ExcelReporter };

File: appium/utils/excel_reporter.py
"""Excel analysis reporter for Appium E2E results."""

from __future__ import annotations

from datetime import datetime
from pathlib import Path

from openpyxl import Workbook
from openpyxl.styles import Alignment, Font, PatternFill


class ExcelReporter:
    def __init__(self, output_path: Path):
        self.output_path = Path(output_path)
        self.rows: list[dict] = []

    def add_result(
        self,
        test_id: str,
        module: str,
        scenario: str,
        status: str,
        duration_ms: float,
        details: str = "",
    ) -> None:
        self.rows.append(
            {
                "test_id": test_id,
                "module": module,
                "scenario": scenario,
                "status": status.upper(),
                "duration_ms": round(duration_ms, 1),
                "details": details,
                "executed_at": datetime.utcnow().strftime("%Y-%m-%d %H:%M:%S UTC"),
            }
        )

    def write(self) -> Path:
        wb = Workbook()

        summary = wb.active
        summary.title = "Summary"
        passed = sum(1 for row in self.rows if row["status"] == "PASSED")
        failed = sum(1 for row in self.rows if row["status"] == "FAILED")
        total = len(self.rows)
        pass_rate = (passed / total * 100.0) if total else 0.0

        summary_headers = ["Metric", "Value"]
        summary.append(summary_headers)
        summary.append(["Total Test Cases", total])
        summary.append(["Passed", passed])
        summary.append(["Failed", failed])
        summary.append(["Pass Rate (%)", round(pass_rate, 2)])
        summary.append(["Generated At", datetime.utcnow().isoformat() + "Z"])
        summary.append(["Framework", "Appium Android E2E + Excel Analysis"])
        self._style_header(summary)

        details = wb.create_sheet("Test Cases")
        details.append(
            [
                "Test ID",
                "Module",
                "Scenario",
                "Status",
                "Duration (ms)",
                "Details",
                "Executed At",
            ]
        )
        for row in self.rows:
            details.append(
                [
                    row["test_id"],
                    row["module"],
                    row["scenario"],
                    row["status"],
                    row["duration_ms"],
                    row["details"],
                    row["executed_at"],
                ]
            )
        self._style_header(details)
        self._color_status(details)

        analysis = wb.create_sheet("Excel Analysis")
        analysis.append(["Module", "Total", "Passed", "Failed", "Pass Rate (%)"])
        modules: dict[str, dict[str, int]] = {}
        for row in self.rows:
            bucket = modules.setdefault(
                row["module"], {"total": 0, "passed": 0, "failed": 0}
            )
            bucket["total"] += 1
            if row["status"] == "PASSED":
                bucket["passed"] += 1
            else:
                bucket["failed"] += 1
        for module, stats in sorted(modules.items()):
            rate = (
                stats["passed"] / stats["total"] * 100.0 if stats["total"] else 0.0
            )
            analysis.append(
                [
                    module,
                    stats["total"],
                    stats["passed"],
                    stats["failed"],
                    round(rate, 2),
                ]
            )
        self._style_header(analysis)

        self.output_path.parent.mkdir(parents=True, exist_ok=True)
        wb.save(self.output_path)
        return self.output_path

    @staticmethod
    def _style_header(sheet) -> None:
        fill = PatternFill("solid", fgColor="1F4E79")
        font = Font(color="FFFFFF", bold=True)
        for cell in sheet[1]:
            cell.fill = fill
            cell.font = font
            cell.alignment = Alignment(horizontal="center")

    @staticmethod
    def _color_status(sheet) -> None:
        green = PatternFill("solid", fgColor="C6EFCE")
        red = PatternFill("solid", fgColor="FFC7CE")
        for row in sheet.iter_rows(min_row=2, min_col=4, max_col=4):
            cell = row[0]
            if str(cell.value).upper() == "PASSED":
                cell.fill = green
            elif str(cell.value).upper() == "FAILED":
                cell.fill = red
